python/pyspark/sql/readwriter.py

```diff
@@ -563,6 +563,60 @@ def partitionBy(self, *cols):
         self._jwrite = self._jwrite.partitionBy(_to_seq(self._spark._sc, cols))
         return self
 
+    @since(2.3)
+    def bucketBy(self, numBuckets, *cols):
+        """Buckets the output by the given columns on the file system.
+
+        :param numBuckets: the number of buckets to save
+        :param cols: name of columns
+
+        .. note:: Applicable for file-based data sources in combination with
+                  :py:meth:`DataFrameWriter.saveAsTable`.
+
+        >>> (df.write.format('parquet')
+        ...     .bucketBy(100, 'year', 'month')
+        ...     .mode("overwrite")
+        ...     .saveAsTable('bucketed_table'))
+        """
+        if len(cols) == 1 and isinstance(cols[0], (list, tuple)):
+            cols = cols[0]
+
+        if not isinstance(numBuckets, int):
+            raise TypeError("numBuckets should be an int, got {0}.".format(type(numBuckets)))
+
+        if not all(isinstance(c, basestring) for c in cols):
+            raise TypeError("cols argument should be a string or a sequence of strings.")
```

Comment:
Good point. We can support arbitrary `Iterable[str]` though. 

```python
if len(cols) == 1 and isinstance(cols[0], collections.abc.Iterable):
    cols = list(cols[0])
```

Caveat is, we don't allow this anywhere else.